File: sdk/src/main/java/com/meniga/sdk/converters/MenigaRedemptionConverter.java
package com.meniga.sdk.converters;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import com.meniga.sdk.helpers.GsonProvider;
import com.meniga.sdk.models.offers.redemptions.MenigaRedemptions;
import com.meniga.sdk.models.offers.redemptions.MenigaScheduledReimbursement;

import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;
import java.util.List;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class MenigaRedemptionConverter extends MenigaConverter {

	@Override
	public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
		Type typeOfRedemptions = new TypeToken<MenigaRedemptions>() {
		}.getType();

		if (typeOfRedemptions.equals(type)) {
			return new Converter<ResponseBody, MenigaRedemptions>() {
				@Override
				public MenigaRedemptions convert(ResponseBody resBody) throws IOException {
					InputStreamReader isr = new InputStreamReader(resBody.byteStream());
					JsonElement element;
					try {
						element = new JsonParser().parse(isr);
					} finally {
						isr.close();
					}


					Gson gson = GsonProvider.getGsonBuilder();
					MenigaRedemptions page = gson.fromJson(getAsArray(element), MenigaRedemptions.class);

					JsonObject object = element.getAsJsonObject();

					JsonObject meta = object.getAsJsonObject("meta");

					List<MenigaScheduledReimbursement> cashBackRaw = gson.fromJson(meta.getAsJsonArray("scheduledReimbursements"), new TypeToken<List<MenigaScheduledReimbursement>>() {
					}.getType());
					MenigaRedemptions main = MenigaConverter.mergeMeta(gson, page, meta);

					main.setScheduledReimbursement(cashBackRaw);
					return main;
				}
			};
		}
		return null;
	}

	@Override
	public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
		return null;
	}
}

File: sdk/src/main/java/com/meniga/sdk/converters/MenigaRedemptionConverter.kt
package com.meniga.sdk.converters

import com.google.gson.Gson
import com.meniga.sdk.helpers.type
import com.meniga.sdk.models.offers.redemptions.MenigaRedemptions
import com.meniga.sdk.models.offers.redemptions.MenigaScheduledReimbursement
import okhttp3.ResponseBody
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
class MenigaRedemptionConverter(private val gson: Gson) : MenigaConverter() {
    override fun responseBodyConverter(type: Type?, annotations: Array<Annotation>?, retrofit: Retrofit?): Converter<ResponseBody, *>? {
        return when (type) {
            type<MenigaRedemptions>() -> Converter<ResponseBody, MenigaRedemptions> { resBody ->
                val (data, meta) = MenigaConverter.getAsArrayApiResponse(resBody.byteStream())

                var menigaRedemptions = gson.fromJson(data, MenigaRedemptions::class.java)

                val cashBackRaw = meta?.let {
                    gson.fromJson<List<MenigaScheduledReimbursement>>(it.getAsJsonArray("scheduledReimbursements"), type<List<MenigaScheduledReimbursement>>())
                }
                meta?.let {
                    menigaRedemptions = MenigaConverter.mergeMeta(gson, menigaRedemptions, it)
                }

                menigaRedemptions.setScheduledReimbursement(cashBackRaw)
                menigaRedemptions
            }
            else -> null
        }
    }
}
